### PR TITLE
[feature] Show weapon skills on Weapon Masters

### LINKED ISSUE #5390: Add weapons to weapon trainers
## Description

It would be great if the weapon trainers also show the weapons they train. [Here is a list](https://www.wowhead.com/classic/guide/classic-wow-weapon-skills#weapon-master-locations)

### PR DESCRIPTION
**Pull request by Sommos**
## Issue references

Fixes #5390

## TODOs

- [x] Add translations for weapon skills
- [X] Make sure the line breaks are working
- [x] Add TBC Weapon Master
- [X] Share screenshots

## Proposed changes

- ADDED: Weapon Skills to Weapon Trainer NPCs to show on map and hover.

image
image

### CODE REVIEW

**Comment by anon1231823** [comment]
Any chance it will be able to automatically localize the weapon skill names?

**Comment by BreakBB** [comment]
> Any chance it will be able to automatically localize the weapon skill names?

Yes, that shouldn't be a problem at all.

**Comment by Sommos** [comment]
Below are a couple of examples of this running on SoD.
image
image

**Comment by Cabro** [comment]
There are a few more weapon masters added with TBC in Silvermoon City, Eversong Woods and The Exodar. We might need to add a check for isTBC and add those NPCs also:

https://www.wowhead.com/wotlk/npc=16621/ileda - silvermoon city
https://www.wowhead.com/wotlk/npc=16773/handiir - the exodar
https://www.wowhead.com/wotlk/npc=17005/duelist-larenis - eversong woods

Following logic, there MIGHT be one more for draenei in their starting zones, but from a quick search i can't find it.

**Comment by Sommos** [comment]
> There are a few more weapon masters added with TBC in Silvermoon City, Eversong Woods and The Exodar. We might need to add a check for isTBC+ and add those NPCs also:
> 
> https://www.wowhead.com/wotlk/npc=16621/ileda - silvermoon city https://www.wowhead.com/wotlk/npc=16773/handiir - the exodar https://www.wowhead.com/wotlk/npc=17005/duelist-larenis - eversong woods
> 
> Following logic, there MIGHT be one more for draenei in their starting zones, but from a quick search i can't find it.

Yeah, it seems that those are all of them, and although logically there would be another on Azuremyst Isle, there is not one there, and not one on Bloodmyst Isle either.

**Comment by Sommos** [comment]
I think we'll probably want to move `weaponMasterSkills` out of this function, so it isn't redefined every time. 

If we do that, we should make a `appendSkillsToTitle` function that can be called to make this cleaner.

**Comment by Sommos** [comment]
> Probably don't need the "Trains the following skills" as it will need to be translated. I think if you just list the weapon skills everyone will know.

Removed this now :)

File: Modules/Map/QuestieMap.lua [hunk 1]
```diff
@@ -359,6 +379,21 @@ function QuestieMap:ShowNPC(npcID, icon, scale, title, body, disableShiftToRemov
     data.IsObjectiveNote = false
     data.ManualTooltipData = {}
     data.ManualTooltipData.Title = title or (npc.name .. " (" .. l10n("NPC") .. ")")
+    local isTBC = Questie.IsTBC
+    local isWotlk = Questie.IsWotlk
```
Comment (BreakBB) [inline Modules/Map/QuestieMap.lua:383]: It's okay to inline these

File: Modules/Map/QuestieMap.lua [hunk 2]
```diff
@@ -332,6 +332,26 @@ function QuestieMap.ProcessQueue()
     end
 end
 
+local weaponMasterSkills = {
```
Comment (BreakBB) [inline Modules/Map/QuestieMap.lua:335]: As you already said: This new logic should be placed somewhere else. QuestieMap seems the wrong place for it. Maybe a small new module is fine.

**Review by BreakBB — CHANGES_REQUESTED** [review]

File: Localization/Translations/WeaponSkills.lua [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+        ["enUS"] = true,
+        ["zhCN"] = "单手剑",
+        ["zhTW"] = "單手劍",
+        ["esES"] = "Espadas de una mano",
+        ["frFR"] = "Épées à une main",
+    },
+    ["One-Handed Maces"] = {
+        ["ptBR"] = "Maças de Uma Mão",
+        ["ruRU"] = "Одноручные дробящие",
+        ["deDE"] = "Einhandschlägel",
```
[the remark this answers is not in the document]
Comment (Sommos) [reply Localization/Translations/WeaponSkills.lua]: Fixed in: https://github.com/Questie/Questie/pull/5568/commits/72357768841ac8ed07c32531114c439a6d32fa0f

File: Localization/Translations/WeaponSkills.lua [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+        ["enUS"] = true,
+        ["zhCN"] = "双手剑",
+        ["zhTW"] = "雙手劍",
+        ["esES"] = "Espadas de dos manos",
+        ["frFR"] = "Épées à deux mains",
+    },
+    ["Two-Handed Maces"] = {
+        ["ptBR"] = "Maças de Duas Mãos",
+        ["ruRU"] = "Двуручные дробящие",
+        ["deDE"] = "Zweihandschlägel",
```
[the remark this answers is not in the document]
Comment (Sommos) [reply Localization/Translations/WeaponSkills.lua]: Fixed in: https://github.com/Questie/Questie/pull/5568/commits/72357768841ac8ed07c32531114c439a6d32fa0f

File: Localization/Translations/WeaponSkills.lua [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        ["enUS"] = true,
+        ["zhCN"] = "双手锤",
+        ["zhTW"] = "雙手錘",
+        ["esES"] = "Mazas de dos manos",
+        ["frFR"] = "Masses à deux mains",
+    },
+    ["Two-Handed Axes"] = {
+        ["ptBR"] = "Machados de Duas Mãos",
+        ["ruRU"] = "Двуручные топоры",
+        ["deDE"] = "Zweihandschwerter",
```
[the remark this answers is not in the document]
Comment (Sommos) [reply Localization/Translations/WeaponSkills.lua]: Fixed in: https://github.com/Questie/Questie/pull/5568/commits/72357768841ac8ed07c32531114c439a6d32fa0f

File: Localization/Translations/WeaponSkills.lua [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+        ["enUS"] = true,
+        ["zhCN"] = "单手锤",
+        ["zhTW"] = "單手錘",
+        ["esES"] = "Mazas de una mano",
+        ["frFR"] = "Masse à une main",
+    },
+    ["One-Handed Axes"] = {
+        ["ptBR"] = "Machados de Uma Mão",
+        ["ruRU"] = "Одноручные топоры",
+        ["deDE"] = "Einhandschwerter",
```
[the remark this answers is not in the document]
Comment (Sommos) [reply Localization/Translations/WeaponSkills.lua]: Fixed in: https://github.com/Questie/Questie/pull/5568/commits/72357768841ac8ed07c32531114c439a6d32fa0f

File: Modules/Map/QuestieMap.lua (hunk 1 above)
Comment (Sommos) [reply Modules/Map/QuestieMap.lua]: Fixed in: https://github.com/Questie/Questie/pull/5568/commits/30f291278b358ebd6c31d50da8fce376b3d2dcbd

File: Modules/Map/QuestieMap.lua (hunk 2 above)
Comment (Sommos) [reply Modules/Map/QuestieMap.lua]: Fixed in: https://github.com/Questie/Questie/pull/5568/commits/e21f912fd598b3550179a024653b37afa5335831

File: Modules/Map/WeaponMasterSkills.lua [hunk 7]
```diff
@@ -0,0 +1,23 @@
+local WeaponMasterSkills = {}
```
Comment (BreakBB) [inline Modules/Map/WeaponMasterSkills.lua:1]: Please don't do this 🙈 

Checkout other "modules" which are using the `QuestieLoader` global to create and import modules.

File: Modules/Map/WeaponMasterSkills.lua [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+    [11867] = {"Crossbows", "Daggers", "One-Handed Swords", "Polearms", "Staves", "Two-Handed Swords"},
+    [13084] = {"Crossbows", "Daggers", "Thrown"},
+    -- Horde Weapon Trainers
+    [2704] = {"Bows", "One-Handed Axes", "Staves", "Thrown", "Two-Handed Axes"},
+    [11868] = {"Bows", "Daggers", "Fist Weapons", "One-Handed Axes", "Thrown", "Two-Handed Axes"},
+    [11869] = {"Guns", "One-Handed Maces", "Staves", "Two-Handed Maces"},
+    [11870] = {"Crossbows", "Daggers", "One-Handed Swords", "Polearms", "Two-Handed Swords"},
+}
+
+function WeaponMasterSkills.appendSkillsToTitle(title, skills)
```
Comment (BreakBB) [inline Modules/Map/WeaponMasterSkills.lua:16]: Function names should be upper case.

File: Modules/Map/QuestieMap.lua [hunk 9]
```diff
@@ -378,17 +358,18 @@ function QuestieMap:ShowNPC(npcID, icon, scale, title, body, disableShiftToRemov
...
     if Questie.IsTBC or Questie.IsWotlk then
         -- Blood Elf Starting Area Weapon Trainers
-        weaponMasterSkills[16621] = {"Bows", "Daggers", "One-Handed Swords", "Polearms", "Thrown", "Two-Handed Swords"}
-        weaponMasterSkills[17005] = {"Bows", "Daggers", "One-Handed Swords", "Polearms", "Thrown", "Two-Handed Swords"}
+        WeaponMasterSkills.data[16621] = {"Bows", "Daggers", "One-Handed Swords", "Polearms", "Thrown", "Two-Handed Swords"}
+        WeaponMasterSkills.data[17005] = {"Bows", "Daggers", "One-Handed Swords", "Polearms", "Thrown", "Two-Handed Swords"}
         -- Draenei Starting Area Weapon Trainers
-        weaponMasterSkills[16773] = {"Crossbows", "Daggers", "One-Handed Maces", "One-Handed Swords", "Two-Handed Maces", "Two-Handed Swords"}
+        WeaponMasterSkills.data[16773] = {"Crossbows", "Daggers", "One-Handed Maces", "One-Handed Swords", "Two-Handed Maces", "Two-Handed Swords"}
     end
```
Comment (BreakBB) [inline Modules/Map/QuestieMap.lua:368]: This should also be moved to the `WeaponMasterSkill` module.

File: Modules/Map/QuestieMap.lua [hunk 10]
```diff
@@ -378,17 +358,18 @@ function QuestieMap:ShowNPC(npcID, icon, scale, title, body, disableShiftToRemov
...
     if npc.subName == "Weapon Master" then
-        local skills = weaponMasterSkills[data.id]
+        local skills = WeaponMasterSkills.data[data.id]
         if skills then
-            data.ManualTooltipData.Title = appendSkillsToTitle(data.ManualTooltipData.Title, skills)
+            data.ManualTooltipData.Title = WeaponMasterSkills.appendSkillsToTitle(data.ManualTooltipData.Title, skills)
         end
```
Comment (BreakBB) [inline Modules/Map/QuestieMap.lua:373]: We should remove the `if` int this place and simply call `WeaponMasterSkills.AppendSkillsToTitle`. If we don't have skills for a specific NPC then log it on warning level and return an empty string (aka nothing is appended).

File: Modules/Map/QuestieMap.lua [hunk 11]
```diff
@@ -328,27 +328,7 @@ function QuestieMap.ProcessQueue()
...
-    [11868] = {"Bows", "Daggers", "Fist Weapons", "One-Handed Axes", "Thrown", "Two-Handed Axes"},
-    [11869] = {"Guns", "One-Handed Maces", "Staves", "Two-Handed Maces"},
-    [11870] = {"Crossbows", "Daggers", "One-Handed Swords", "Polearms", "Two-Handed Swords"},
-}
-local function appendSkillsToTitle(title, skills)
-    for _, skill in ipairs(skills) do
-        title = title .. "\n - " .. l10n(skill)
-    end
-    return title
+        end
```
Comment (BreakBB) [inline Modules/Map/QuestieMap.lua:331]: You removed one `end` too much, so the entire code is broken atm. Did you test this?

**Review by BreakBB — CHANGES_REQUESTED** [review]

File: Modules/Map/WeaponMasterSkills.lua (hunk 8 above)
Comment (Sommos) [reply Modules/Map/WeaponMasterSkills.lua]: Fixed in: https://github.com/Questie/Questie/pull/5568/commits/2beca70ea6220a1102f0b8fc9057a3b3afd992c4

File: Modules/Map/QuestieMap.lua (hunk 9 above)
Comment (Sommos) [reply Modules/Map/QuestieMap.lua]: Fixed in: https://github.com/Questie/Questie/pull/5568/commits/2beca70ea6220a1102f0b8fc9057a3b3afd992c4

File: Modules/Map/WeaponMasterSkills.lua (hunk 7 above)
Comment (Sommos) [reply Modules/Map/WeaponMasterSkills.lua]: This should now be fixed.

File: Modules/Map/QuestieMap.lua (hunk 11 above)
Comment (Sommos) [reply Modules/Map/QuestieMap.lua]: For some reason, this was not the case locally. So very unsure how this has gotten in there. Anyways, it's now fixed.

File: Modules/Map/QuestieMap.lua [hunk 12]
```diff
@@ -329,7 +329,6 @@ function QuestieMap.ProcessQueue()
         if minimapDrawCall[2]._needsUnload then
             minimapDrawCall[2]:Unload()
         end
-    end
```
Comment (BreakBB) [inline Modules/Map/QuestieMap.lua:332]: This `end` still needs to be added back.

File: Modules/Map/WeaponMasterSkills.lua [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+end
+
+function WeaponMasterSkills.AppendSkillsToTitle(title, skills)
+    for _, skill in ipairs(skills) do
+        title = title .. "\n - " .. l10n(skill)
+    end
+    return title
+end
+
+return WeaponMasterSkills
```
Comment (BreakBB) [inline Modules/Map/WeaponMasterSkills.lua:35]: This is not needed

```suggestion
```

File: Modules/Map/QuestieMap.lua [hunk 14]
```diff
@@ -359,6 +358,13 @@ function QuestieMap:ShowNPC(npcID, icon, scale, title, body, disableShiftToRemov
     data.IsObjectiveNote = false
     data.ManualTooltipData = {}
     data.ManualTooltipData.Title = title or (npc.name .. " (" .. l10n("NPC") .. ")")
+    local WeaponMasterSkills = require("WeaponMasterSkills")
```
Comment (BreakBB) [inline Modules/Map/QuestieMap.lua:361]: Use `QuestieLoader:ImportModule` on top of the file instead of inline `require`.

**Review by BreakBB — CHANGES_REQUESTED** [review]

File: Modules/Map/QuestieMap.lua (hunk 14 above)
Comment (Sommos) [reply Modules/Map/QuestieMap.lua]: Fixed in: https://github.com/Questie/Questie/pull/5568/commits/9e516ebf1afbac78459a49a1d8fe7b26c0f1dcf1

File: Modules/Map/WeaponMasterSkills.lua (hunk 13 above)
Comment (Sommos) [reply Modules/Map/WeaponMasterSkills.lua]: Fixed in: https://github.com/Questie/Questie/pull/5568/commits/9e516ebf1afbac78459a49a1d8fe7b26c0f1dcf1

File: Modules/Map/QuestieMap.lua (hunk 12 above)
Comment (Sommos) [reply Modules/Map/QuestieMap.lua]: I'm not totally sure why this was not fixed last time. Anyways, it should be fixed now.

File: Modules/Map/QuestieMap.lua [hunk 15]
```diff
@@ -359,6 +361,11 @@ function QuestieMap:ShowNPC(npcID, icon, scale, title, body, disableShiftToRemov
     data.IsObjectiveNote = false
     data.ManualTooltipData = {}
     data.ManualTooltipData.Title = title or (npc.name .. " (" .. l10n("NPC") .. ")")
+    local skills = WeaponMasterSkills.data[data.id]
+    if not skills then
+        Questie:Debug(Questie.DEBUG_DEVELOP, "", data.id)
+    end
+    data.ManualTooltipData.Title = WeaponMasterSkills:AppendSkillsToTitle(data.ManualTooltipData.Title, skills or {})
```
Comment (BreakBB) [inline Modules/Map/QuestieMap.lua:368]:
```suggestion
    local baseTitle = title or (npc.name .. " (" .. l10n("NPC") .. ")")
    data.ManualTooltipData.Title = WeaponMasterSkills.AppendSkillsToTitle(baseTitle, data.id)
```

File: Modules/Map/WeaponMasterSkills.lua [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+
+function WeaponMasterSkills.AppendSkillsToTitle(title, skills)
+    if #skills == 0 then
+        return title
+    end
+    for _, skill in ipairs(skills) do
+        title = title .. "\n - " .. l10n(skill)
+    end
+    return title
+end
```
Comment (BreakBB) [inline Modules/Map/WeaponMasterSkills.lua:36]:
```suggestion
function WeaponMasterSkills.AppendSkillsToTitle(title, npcId)
    local skills = WeaponMasterSkills.data[npcId]
    if (not skills) then
        return title
    end

    for _, skill in ipairs(skills) do
        title = title .. "\n - " .. l10n(skill)
    end
    return title
end
```

**Review by BreakBB — COMMENTED** [review]
I suggest a few things in the comments:

1. Omit the debug print (I know I "asked" for it, but it's not as helpful as I expected)
2. Move the existence check of skills into the append function, so only the function is called outside of the new module
3. Fixed the call of the function (`module:func` vs `module.func`)

**Review by BreakBB — CHANGES_REQUESTED** [review]
One thing that is missing, is to actually load the `WeaponMasterSkills` in the TOC files. Otherwise it will won't work.

**Comment by BreakBB** [comment]
Please also add it to the BCC and WotLK TOC file. Otherwise it will break on WotLK.

**Pull request merged**